Allow pagination of mobile OrgUnit API

bmonjoie opened the pull request:
We received a complaint that some devices cannot synchronize OrgUnits when they are too many.
As we couldn't reproduce, the assumptions are:
 - The list is too big in memory
 - The filtering that is done to remove the `parent_id` value of children for which the parent hasn't been returned is too consuming.

This PR allows 2 new parameters: `page_size`and `page`, to process the OrgUnit in smaller chunks.
Since we won't be able to remove the `parent_id` of children without parents with only part of the data, the removal is now done in the backend.

Therefore, this addresses both assumptions while still being backward compatible.

Related JIRA tickets: IA-1914

## Self proofreading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented
- [ ] Are my typescript files well typed
- [ ] New translations have been added or updated if new string have been introduced in the frontend
- [ ] My migrations file are included
- [X] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

The `/api/mobile/orgunits` endpoint is now capable of returning a paginated list.
The `parent_id` value of OrgUnits with a parent that has a `REJECTED` status is now set to `None`.

## How to test

Call `/api/mobile/orgunits?app_id=...` **before** applying the changes.
Call `/api/mobile/orgunits?app_id=...` **after** applying the changes.
The results should be the same

Then, call `/api/mobile/orgunits?app_id=...` with a `page_size` and `page` parameters and ensure the right chunk of the data has been returned.

bmonjoie committed:
Allow pagination of mobile OrgUnit API

IA-1914

bmonjoie commented on iaso/models/org_unit.py:
```diff
@@ -371,6 +371,13 @@ def as_dict_for_mobile(self):
+            if self.parent_id is None or self.parent.validation_status == OrgUnit.VALIDATION_VALID
+            else (self.parent.get_valid_parent_id(default_to_none) if default_to_none is not True else None)
+        )
+
```
@olethanh Even though I have added `.prefetch_related("parent")` to the `queryset`, I'm afraid this could result in a lot of queries to the database.
Could you advise if this is OK and, if not, what would be a better way?

olethanh replied:
you can do a .prefetch_related("parent").prefech_related("parent_parent").. etc if you do 4 level it should be enough

bmonjoie committed:
Add ordering

IA-1914

bmonjoie committed:
Try to fix mypy error

bmonjoie committed:
Adding uuid to the mobile response

The mobile application has currently no way to know if a custom org unit
should be updated or removed because it was approved.
Adding this parameter will allow the mobile application to normalize the data.

madewulf commented on iaso/api/mobile/org_units.py:
```diff
@@ -3,14 +3,22 @@
+    page_size_query_param = PAGE_SIZE
+    page_query_param = PAGE
+    page_size = None  # None to disable pagination by default.
+
```
Everywhere else, the paginations parameter is limit instead of page_size, I think. It would be better to stick with that

bmonjoie replied:
I first put `limit` but I saw that you put `page_size` on entities so I used that to stay consistent:
![image](https://user-images.githubusercontent.com/564099/223986719-42a6a6df-eec4-4cd2-bbc9-67263d6812d0.png)

madewulf replied:
That's a mistake. We should stick to limit. Do you use that parameter already on entities?

bmonjoie replied:
Actually, the code on mobile passes `limit` :sweat_smile: , I'll change it in both places

madewulf replied:
👍

madewulf commented on iaso/api/mobile/org_units.py:
```diff
@@ -32,52 +40,66 @@ def has_object_permission(self, request, view, obj):
+        page = pagination.paginate_queryset(queryset, request)
+        page_number = 1
+        page_size = None
+        if page:
```
This works, and you probably found samples like these in the older APIs, but pagination can be fully handled by Django Rest Framework with way less code

madewulf replied:
But maybe this is done in this way to stay backward compatible?

bmonjoie replied:
I know, and I tried, but for the Django Rest Framework to handle the pagination fully, we have to use a serializer. Since the serialization is done through an `as_dict_for_mobile()` method, I couldn't use a serializer without redoing it all.

madewulf replied:
I think that would be a very small serializer as it does not follow any foreign key
<img width="881" alt="CleanShot 2023-03-09 at 11 09 35@2x" src="https://user-images.githubusercontent.com/185797/223989814-eb701c6c-22f2-42cd-b137-923ddeb2568b.png">
but if some older apps are expecting results without pagination, then, it's probably painful to get right in the DRF (that will paginate everything by default, which is probably a sane approach)

madewulf replied:
All those as_dict() method were just me being ignorant of good practices of the DRF and ideally should disappear, FYI.

madewulf commented on iaso/tests/api/test_mobile_orgunits.py:
```diff
@@ -0,0 +1,127 @@
+from iaso.api.query_params import APP_ID, PAGE_SIZE, PAGE
```
Was there no test for this mobile org unit api before?

bmonjoie replied:
None that I could find

madewulf replied:
Could you add a test for the non paginated version then? It should be fast with what you have written already. Thanks!

bmonjoie replied:
That's literally my first test :thinking:

bmonjoie committed:
Change query param "page_size" to "limit"

bmonjoie committed:
Use serializer and Django pagination tools

bmonjoie committed:
Revert changes made to `as_dict_for_mobile`

mathvdh commented on iaso/api/mobile/org_units.py:
```diff
@@ -1,16 +1,91 @@
 from iaso.models import OrgUnit, Project
 
 
+class MobileOrgUnitsSetPagination(PageNumberPagination):
```
Why are we not subclassing the Paginator from `/iaso/api/common.py` ? (Just a question ..)

bmonjoie replied:
Because it's not the one that was used by @madewulf in [`/iaso/api/mobile/entity.py`](/BLSQ/iaso/pull/425/files#diff-5098f37ac9f52301b6841b2fe96a74bf556dcca106bcb9c0fdf4659985762bce). Should I change to the one in `/iaso/api/common.py`?

bmonjoie replied:
I have changed to use the one in `/iaso/api/common.py`

bmonjoie committed:
Use `Paginator` from `iaso.api.common`

olethanh committed:
IA-1979 include geojson

rework pagination  a bit

olethanh committed:
Remove location

bmonjoie committed:
Fix type bug and add doc
